Mostly accurate jumping from diff to file

kaste opened the pull request:
Jumping from diff to file really was super sloppy implemented. It almost never jumped to the correct position. 

Implements mostly accurate jumping from diff to file by actually
parsing the hunk and recounting its lines.

For lines which are present on the b side (the view we jump to) this
is accurate of course. For deleted lines ('-') we, mildly arbitrary,
have to choose a good enough line.

If the user is not within a hunk, jump to the first visible line with content.

asfaltboy commented on core/commands/diff.py:
```diff
@@ -18,6 +18,17 @@
 from ...common import util
 
 
+if False:
```
I might be wrong, but shouldn't this be `MYPY = False; if MYPY:` ? Otherwise, would it still be used correctly by mypy?

Ref: https://github.com/python/mypy/issues/3216#issuecomment-296318286

kaste replied:
I've seen both and both do work with mypy. If Guido wants your version I can change it but both do look idiosyncratic.

asfaltboy replied:
Good to know the `if False` works too, agree it's super weird

asfaltboy commented on core/commands/diff.py:
```diff
@@ -468,54 +484,198 @@ class GsDiffOpenFileAtHunkCommand(TextCommand, GitCommand):
+        if not filename:
+            return None
+
+        return filename, row, max(col, 1)
```
another suggestion, how about we return a named tuple instead of regular tuple here? This would make it easier to refer to each jump position item's arguments without indexing / unpacking, and can still be passed as "star args" if want?

kaste commented:
Overall, the code isn't very clear esp. the `col` thing. I think the problem here is when exactly col (and row) are zero based or 1-based. Usually iirc `rowcol` is zero based but `window.open_file` expects 1-based rowcols. git's diff is 1 based for the line (and doesn't provide a col). No I wonder why I often return `b, 0` etc. when I do a `max(col, 1)` later anyway. 🤔 

Now, I hope the tests are good. 'Mostly accurate' is iirc only when there are only deleted lines ('-') in a hunk it is a bit arbitrary where to jump to. Should we jump to the next or the previous line? Currently I always jump to col 0 in that case as well. Maybe that should change.

kaste commented:
Also, 'mostly accurate' are the symmetric '+-' blocks like

```
- foo
- foox
+ bar
+ barz
```

If you're on the 'foox' line it still jumps to the first '+' line, here 'bar'. Ideally maybe it would jump to the 'barz' line. But then it is not enough to just recount the lines, I also must group them to find such '+-' blocks.

kaste commented:
I actually rewrote this already without mention it here. 

Two notes: 

1. I extracted `parse_diff_in_view` and `head_and_hunk_for_pt` almost verbatim from `GsDiffStageOrResetHunkCommand`. These are also extracted and used in #1120. In a follow up PR I would refactor `GsDiffStageOrResetHunkCommand` to use these two functions instead. 

2. I don't want to introduce the namedtuple here. The data structure is private and only used in the once-per-filename filter. I actually can imagine to support selections here as well, t.i. if you select a word in the diff it will select the same word in the original file, which would make the namedtuple moot immediately.

kaste commented:
I learned in the meantime that 'show_commit' uses this functionality here as well, so there is a small fixup for that.

It was (maybe) a bit confusing to jump to the first changed line if the cursor is on the hunk header, so we jump to the first visible line instead.

kaste commented:
This is basically a fix for a currently broken feature. PTAL!

stoivo commented on core/commands/diff.py:
```diff
@@ -573,54 +576,167 @@ class GsDiffOpenFileAtHunkCommand(TextCommand, GitCommand):
+    # type: (sublime.View, int, int) -> Tuple[int, int]
+    """Return rowcol of given pt relative to hunk start"""
+    head_row, _ = view.rowcol(hunk_start)
+    row_on_pt, col = view.rowcol(pt)
```
rename to `pt_row`

stoivo commented on core/commands/diff.py:
```diff
@@ -573,54 +577,178 @@ class GsDiffOpenFileAtHunkCommand(TextCommand, GitCommand):
+    if not match:
+        return None
+
+    b = int(match.group(2))
```
b, we can use a better name.

stoivo replied:
Not neccessary, just a litt confusing when you start

stoivo commented:
You can do the fixup now. When through once tonight and will do once more tomorrow

kaste commented:
Ok, done another rebasing.

stoivo commented on core/commands/diff.py:
```diff
@@ -571,54 +575,178 @@ class GsDiffOpenFileAtHunkCommand(TextCommand, GitCommand):
+                if line.mode in ('+', ' ') and line.text.strip()
+            ),
+            1
+        )
```
As I understand these lines of code. Get me the index of the first line with  + or - and the like text in not blank. Why the not blank check?

kaste replied:
This is basically an arbitrary choice. As I tested this it just felt better if the cursor jumped to a position with code on it. Iirc I added this as a separate commit so it was literally fine tuning the algorithm.

These edge case decisions are why I say mostly accurate jumping. Most of the time it is accurate jumping t.i. fixing the previous buggy behavior. And then I added two arbitrary decisions which I think feel a tiny bit better.

stoivo commented:
a type in this commit message https://github.com/divmain/GitSavvy/pull/1119/commits/273ca5ff44670b0c6577834829c1e8fe5b547257

`Select fist line with content if on a hunk header line ` => `Select first line with content if on a hunk header line`

kaste commented:
Fixed the typo, that was actually the commit where I added the 'not on a blank line' variant.

kaste commented:
🙏
